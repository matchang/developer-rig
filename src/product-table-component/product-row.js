import React from 'react';
import PropTypes from 'prop-types';

export const ProductRow = (props) => {
  const validationErrors = props.product.validationErrors || {};
  const invalid = Object.keys(validationErrors).length > 0;

  return (
    <div className="product-table__row">
      <div className="text-col">
        <input type="text" 
          className={validationErrors.displayName ? "form__input invalid" : "form__input"}
          name="displayName"
          required
          maxLength="255"
          value={props.product.displayName}
          onChange={props.handleValueChange}
        />
        {invalid &&
          <p className="invalid-hint">
            {validationErrors.displayName}
          </p>
        }
      </div>
      <div className="text-col">
        <input type="text"
          className={validationErrors.sku ? "form__input invalid" : "form__input"}
          name="sku"
          required
          maxLength="255"
          value={props.product.sku}
          onChange={props.handleValueChange}
        />
        {invalid &&
          <p className="invalid-hint">
            {validationErrors.sku}
          </p>
        }
      </div>
      <div className="text-col">
        <input type="number"
          className={validationErrors.amount ? "form__input invalid" : "form__input"}
          name="amount"
          required
          min="1" max="10000"
          value={props.product.amount}
          onChange={props.handleValueChange}
        />
        {invalid &&
          <p className="invalid-hint">
            {validationErrors.amount}
          </p>
        }
      </div>
      <div className="select-col">
        <select name="inDevelopment"
            className={validationErrors.inDevelopment ? "form__input invalid" : "form__input"}
            value={props.product.inDevelopment}
            onChange={props.handleValueChange}>
          <option value="true">Yes</option>
          <option value="false">No</option>
        </select>
        {invalid &&
          <p className="invalid-hint">
            {validationErrors.inDevelopment}
          </p>
        }
      </div>
      <div className="select-col">
        <select name="broadcast"
            className={validationErrors.broadcast ? "form__input invalid" : "form__input"}
            value={props.product.broadcast}
            onChange={props.handleValueChange}>
          <option value="true">Yes</option>
          <option value="false">No</option>
        </select>
        {invalid &&
          <p className="invalid-hint">
            {validationErrors.broadcast}
          </p>
        }
      </div>
      <div className="dirty-col">
        {props.product.dirty &&
          <div className="dirty-indicator">
            <svg xmlns='http://www.w3.org/2000/svg' width='30' height='30'>
              <circle cx='5' cy='5' r='3' />
            </svg>
          </div>
        }
        {invalid && <p className="invalid-hint"></p>}
      </div>
    </div>
  );
}

ProductRow.propTypes = {
  product: PropTypes.object.isRequired
}