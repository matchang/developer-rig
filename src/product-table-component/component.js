import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './component.sass';
import { ProductRow } from './product-row';
import { fetchProducts, saveProduct } from '../util/api';

export class ProductTable extends Component {
  constructor(props) {
    super(props);

    this.state = {
      products: [{
        displayName: '',
        sku: '',
        amount: '1',
        inDevelopment: 'true',
        broadcast: 'true',
        validationErrors: {}
      }],
      error: ''
    };
  }

  componentDidMount() {
    fetchProducts(
      'api.twitch.tv',
      this.props.clientId,
      this._handleFetchProductsSuccess.bind(this),
      this._handleFetchProductsError.bind(this)
    );
  }

  handleValueChange(index, event) {
    const value = event.target.value;
    const fieldName = event.target.name;
    const partial = {
      [fieldName]: value,
      dirty: true
    };
    this._updateProduct(index, partial);
  }

  handleAddProductClick(event) {
    this.setState(prevState => {
      const products = [...prevState['products']];
      const product = {
        displayName: 'New Product',
        sku: 'newSKU',
        amount: 1,
        inDevelopment: 'true',
        broadcast: 'true',
        dirty: true
      };
      products.push(product)
      return { products: products };
    });
  }

  handleSaveProductsClick(event) {
    const dirtyProducts = this.state.products.map((p, i) => {
      if (p.dirty) {
        p.saving = true
        saveProduct(
          'api.twitch.tv',
          this.props.clientId,
          this.props.token,
          p,
          this._handleSaveProductSuccess.bind(this, i),
          this._handleSaveProductError.bind(this, i)
        );
      }
      return p;
    });
    this.setState({
      products: dirtyProducts
    });
  }

  render() {
    let disableSaveButton = false;
    const skus = this.state.products.map(p => p.sku);

    const productRows = this.state.products.map((p, i) => {
      const matchingSkus = skus.filter(sku => sku === p.sku);
      if (matchingSkus.length > 1) {
        p.validationErrors = {
          ...p.validationErrors,
          sku: 'SKU must be unique'
        }
      } else if (p.validationErrors.sku === 'SKU must be unique') {
        delete p.validationErrors.sku;
      }

      if (p.validationErrors && Object.keys(p.validationErrors).length > 0) {
        disableSaveButton = true;
      }

      return (
        <ProductRow key={i} product={p} handleValueChange={this.handleValueChange.bind(this, i)} />
      );
    });

    return (
      <div className="product-table">
        {this.state.error &&
          <div className="product-table__error">
            <h4>Error getting products.</h4>
            <p>{this.state.error}</p>
          </div>
        }
        <div className="product-table__header">
          <div className="text-col">Product Name</div>
          <div className="text-col">SKU</div>
          <div className="text-col">Amount (in Bits)</div>
          <div className="select-col">In Development</div>
          <div className="select-col">Broadcast</div>
          <div className="dirty-col"></div>
        </div>
        {productRows}
        <div className="product-table__buttons">
          <button className="product-table__add-button" onClick={this.handleAddProductClick.bind(this)}>
            Add Product
          </button>
          <button className="product-table__save-button"
              onClick={this.handleSaveProductsClick.bind(this)}
              disabled={disableSaveButton}>
            Save All
          </button>
        </div>
      </div>
    );
  }

  _updateProduct(index, partial) {
    this.setState(prevState => {
      const products = prevState.products.map((product, idx) => {
        if (idx === index) {
          let newProduct = {
            ...product,
            ...partial
          };
          newProduct.validationErrors = this._validateProduct(newProduct);
          return newProduct;
        }
        return product;
     });
     return { products: products };
    });
  }

  _validateProduct(product) {
    let validationErrors = {};

    if (!product.displayName) {
      validationErrors.displayName = 'Name must not be empty';
    } else if (product.displayName.length > 255) {
      validationErrors.displayName = 'Name must be less than 256 characters';
    }

    if (!product.sku) {
      validationErrors.sku = 'SKU must not be empty'
    } else if (product.sku.search(/^\S*$/)) {
      validationErrors.sku = 'SKU must not contain any whitespace'
    } else if (product.sku.length > 255) {
      validationErrors.sku = 'SKU must be less than 256 characters';
    }

    if (!product.amount) {
      validationErrors.amount = 'Amount must not be empty'
    } else if (product.amount < 1 || product.amount > 10000) {
      validationErrors.amount = 'Amount must be between 1 and 10,000'
    }

    return validationErrors;
  }

  _handleFetchProductsSuccess(products) {
    this.setState({ products: products });
  }

  _handleFetchProductsError(error) {
    this.setState({ error: error });
  }

  _handleSaveProductSuccess(index) {
    const partial = {
      dirty: false
    };
    this._updateProduct(index, partial);
  }

  _handleSaveProductError(index, error) {
    const partial = {
      error: error
    };
    this._updateProduct(index, partial);
  }
}

ProductTable.propTypes = {
  clientId: PropTypes.string.isRequired,
  token: PropTypes.string.isRequired
}