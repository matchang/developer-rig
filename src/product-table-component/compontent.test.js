import { setupShallowTest } from '../tests/enzyme-util/shallow';
import { ProductTable } from './component';
import { ProductRow } from './product-row';

function mockApiFunctions() {
  const original = require.requireActual('../util/api');
  return {
    ...original,
    fetchProducts: jest.fn(() => { return PRODUCTS; })
  }
}
jest.mock('../util/api', () => mockApiFunctions());
const api = require.requireMock('../util/api');

const CLIENT_ID = 'client-id';
const TOKEN = 'token';
const PRODUCTS = [
  {
    sku: 'test1',
    displayName: 'Test 1',
    amount: '1',
    inDevelopment: 'true',
    broadcast: 'false',
    validationErrors: {}
  },
  {
    sku: 'test2',
    displayName: 'Test 2',
    amount: '100',
    inDevelopment: 'false',
    broadcast: 'false',
    validationErrors: {}
  }
];

describe('<ProductTable />', () => {
  const setupShallow = setupShallowTest(ProductTable, () => ({
    clientId: CLIENT_ID,
    token: TOKEN
  }));

  it('renders correctly', () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it('should fetch products', () => {
    expect(api.fetchProducts).toHaveBeenCalledTimes(1);
  });
});

describe('<ProductRow />', () => {
  const setupShallow = setupShallowTest(ProductRow, () => ({
    product: PRODUCTS[0]
  }));

  it('renders correctly', () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it('shows validation errors', () => {
    const { wrapper } = setupShallow({
      product: {
        ...PRODUCTS[0],
        validationErrors: {
          sku: 'SKU is invalid'
        }
      }
    });
    expect(wrapper.find('input.invalid[name="sku"]')).toHaveLength(1);
    expect(wrapper.find('p.invalid-hint').filterWhere(n => n.text() === 'SKU is invalid')).toHaveLength(1);
  });

  it('shows row as dirty', () => {
    const { wrapper } = setupShallow({
      product: {
        ...PRODUCTS[0],
        dirty: true
      }
    });
    expect(wrapper.find('.dirty-indicator')).toHaveLength(1);
  });
});