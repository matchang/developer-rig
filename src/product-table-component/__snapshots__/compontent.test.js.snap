// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductRow /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    product={
        Object {
            "amount": "1",
            "broadcast": "false",
            "displayName": "Test 1",
            "inDevelopment": "true",
            "sku": "test1",
            "validationErrors": Object {},
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="text-col"
>
          <input
                    className="form__input"
                    maxLength="255"
                    name="displayName"
                    onChange={undefined}
                    required={true}
                    type="text"
                    value="Test 1"
          />
</div>,
        <div
          className="text-col"
>
          <input
                    className="form__input"
                    maxLength="255"
                    name="sku"
                    onChange={undefined}
                    required={true}
                    type="text"
                    value="test1"
          />
</div>,
        <div
          className="text-col"
>
          <input
                    className="form__input"
                    max="10000"
                    min="1"
                    name="amount"
                    onChange={undefined}
                    required={true}
                    type="number"
                    value="1"
          />
</div>,
        <div
          className="select-col"
>
          <select
                    className="form__input"
                    name="inDevelopment"
                    onChange={undefined}
                    value="true"
          >
                    <option
                              value="true"
                    >
                              Yes
                    </option>
                    <option
                              value="false"
                    >
                              No
                    </option>
          </select>
</div>,
        <div
          className="select-col"
>
          <select
                    className="form__input"
                    name="broadcast"
                    onChange={undefined}
                    value="false"
          >
                    <option
                              value="true"
                    >
                              Yes
                    </option>
                    <option
                              value="false"
                    >
                              No
                    </option>
          </select>
</div>,
        <div
          className="dirty-col"
>
          
</div>,
      ],
      "className": "product-table__row",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="form__input"
              maxLength="255"
              name="displayName"
              onChange={undefined}
              required={true}
              type="text"
              value="Test 1"
/>,
            false,
          ],
          "className": "text-col",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form__input",
              "maxLength": "255",
              "name": "displayName",
              "onChange": undefined,
              "required": true,
              "type": "text",
              "value": "Test 1",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          false,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="form__input"
              maxLength="255"
              name="sku"
              onChange={undefined}
              required={true}
              type="text"
              value="test1"
/>,
            false,
          ],
          "className": "text-col",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form__input",
              "maxLength": "255",
              "name": "sku",
              "onChange": undefined,
              "required": true,
              "type": "text",
              "value": "test1",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          false,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="form__input"
              max="10000"
              min="1"
              name="amount"
              onChange={undefined}
              required={true}
              type="number"
              value="1"
/>,
            false,
          ],
          "className": "text-col",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form__input",
              "max": "10000",
              "min": "1",
              "name": "amount",
              "onChange": undefined,
              "required": true,
              "type": "number",
              "value": "1",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          false,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <select
              className="form__input"
              name="inDevelopment"
              onChange={undefined}
              value="true"
>
              <option
                            value="true"
              >
                            Yes
              </option>
              <option
                            value="false"
              >
                            No
              </option>
</select>,
            false,
          ],
          "className": "select-col",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option
                  value="true"
>
                  Yes
</option>,
                <option
                  value="false"
>
                  No
</option>,
              ],
              "className": "form__input",
              "name": "inDevelopment",
              "onChange": undefined,
              "value": "true",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Yes",
                  "value": "true",
                },
                "ref": null,
                "rendered": "Yes",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "No",
                  "value": "false",
                },
                "ref": null,
                "rendered": "No",
                "type": "option",
              },
            ],
            "type": "select",
          },
          false,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <select
              className="form__input"
              name="broadcast"
              onChange={undefined}
              value="false"
>
              <option
                            value="true"
              >
                            Yes
              </option>
              <option
                            value="false"
              >
                            No
              </option>
</select>,
            false,
          ],
          "className": "select-col",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option
                  value="true"
>
                  Yes
</option>,
                <option
                  value="false"
>
                  No
</option>,
              ],
              "className": "form__input",
              "name": "broadcast",
              "onChange": undefined,
              "value": "false",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Yes",
                  "value": "true",
                },
                "ref": null,
                "rendered": "Yes",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "No",
                  "value": "false",
                },
                "ref": null,
                "rendered": "No",
                "type": "option",
              },
            ],
            "type": "select",
          },
          false,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            false,
          ],
          "className": "dirty-col",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          false,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="text-col"
>
            <input
                        className="form__input"
                        maxLength="255"
                        name="displayName"
                        onChange={undefined}
                        required={true}
                        type="text"
                        value="Test 1"
            />
</div>,
          <div
            className="text-col"
>
            <input
                        className="form__input"
                        maxLength="255"
                        name="sku"
                        onChange={undefined}
                        required={true}
                        type="text"
                        value="test1"
            />
</div>,
          <div
            className="text-col"
>
            <input
                        className="form__input"
                        max="10000"
                        min="1"
                        name="amount"
                        onChange={undefined}
                        required={true}
                        type="number"
                        value="1"
            />
</div>,
          <div
            className="select-col"
>
            <select
                        className="form__input"
                        name="inDevelopment"
                        onChange={undefined}
                        value="true"
            >
                        <option
                                    value="true"
                        >
                                    Yes
                        </option>
                        <option
                                    value="false"
                        >
                                    No
                        </option>
            </select>
</div>,
          <div
            className="select-col"
>
            <select
                        className="form__input"
                        name="broadcast"
                        onChange={undefined}
                        value="false"
            >
                        <option
                                    value="true"
                        >
                                    Yes
                        </option>
                        <option
                                    value="false"
                        >
                                    No
                        </option>
            </select>
</div>,
          <div
            className="dirty-col"
>
            
</div>,
        ],
        "className": "product-table__row",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="form__input"
                maxLength="255"
                name="displayName"
                onChange={undefined}
                required={true}
                type="text"
                value="Test 1"
/>,
              false,
            ],
            "className": "text-col",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form__input",
                "maxLength": "255",
                "name": "displayName",
                "onChange": undefined,
                "required": true,
                "type": "text",
                "value": "Test 1",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            false,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="form__input"
                maxLength="255"
                name="sku"
                onChange={undefined}
                required={true}
                type="text"
                value="test1"
/>,
              false,
            ],
            "className": "text-col",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form__input",
                "maxLength": "255",
                "name": "sku",
                "onChange": undefined,
                "required": true,
                "type": "text",
                "value": "test1",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            false,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="form__input"
                max="10000"
                min="1"
                name="amount"
                onChange={undefined}
                required={true}
                type="number"
                value="1"
/>,
              false,
            ],
            "className": "text-col",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form__input",
                "max": "10000",
                "min": "1",
                "name": "amount",
                "onChange": undefined,
                "required": true,
                "type": "number",
                "value": "1",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            false,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <select
                className="form__input"
                name="inDevelopment"
                onChange={undefined}
                value="true"
>
                <option
                                value="true"
                >
                                Yes
                </option>
                <option
                                value="false"
                >
                                No
                </option>
</select>,
              false,
            ],
            "className": "select-col",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <option
                    value="true"
>
                    Yes
</option>,
                  <option
                    value="false"
>
                    No
</option>,
                ],
                "className": "form__input",
                "name": "inDevelopment",
                "onChange": undefined,
                "value": "true",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Yes",
                    "value": "true",
                  },
                  "ref": null,
                  "rendered": "Yes",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "No",
                    "value": "false",
                  },
                  "ref": null,
                  "rendered": "No",
                  "type": "option",
                },
              ],
              "type": "select",
            },
            false,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <select
                className="form__input"
                name="broadcast"
                onChange={undefined}
                value="false"
>
                <option
                                value="true"
                >
                                Yes
                </option>
                <option
                                value="false"
                >
                                No
                </option>
</select>,
              false,
            ],
            "className": "select-col",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <option
                    value="true"
>
                    Yes
</option>,
                  <option
                    value="false"
>
                    No
</option>,
                ],
                "className": "form__input",
                "name": "broadcast",
                "onChange": undefined,
                "value": "false",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Yes",
                    "value": "true",
                  },
                  "ref": null,
                  "rendered": "Yes",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "No",
                    "value": "false",
                  },
                  "ref": null,
                  "rendered": "No",
                  "type": "option",
                },
              ],
              "type": "select",
            },
            false,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              undefined,
              false,
            ],
            "className": "dirty-col",
          },
          "ref": null,
          "rendered": Array [
            undefined,
            false,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<ProductTable /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProductTable
    clientId="client-id"
    token="token"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        "",
        <div
          className="product-table__header"
>
          <div
                    className="text-col"
          >
                    Product Name
          </div>
          <div
                    className="text-col"
          >
                    SKU
          </div>
          <div
                    className="text-col"
          >
                    Amount (in Bits)
          </div>
          <div
                    className="select-col"
          >
                    In Development
          </div>
          <div
                    className="select-col"
          >
                    Broadcast
          </div>
          <div
                    className="dirty-col"
          />
</div>,
        Array [
          <Unknown
            handleValueChange={[Function]}
            product={
                        Object {
                                    "amount": "1",
                                    "broadcast": "true",
                                    "displayName": "",
                                    "inDevelopment": "true",
                                    "sku": "",
                                    "validationErrors": Object {},
                                  }
            }
/>,
        ],
        <div
          className="product-table__buttons"
>
          <button
                    className="product-table__add-button"
                    onClick={[Function]}
          >
                    Add Product
          </button>
          <button
                    className="product-table__save-button"
                    disabled={false}
                    onClick={[Function]}
          >
                    Save All
          </button>
</div>,
      ],
      "className": "product-table",
    },
    "ref": null,
    "rendered": Array [
      "",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="text-col"
>
              Product Name
</div>,
            <div
              className="text-col"
>
              SKU
</div>,
            <div
              className="text-col"
>
              Amount (in Bits)
</div>,
            <div
              className="select-col"
>
              In Development
</div>,
            <div
              className="select-col"
>
              Broadcast
</div>,
            <div
              className="dirty-col"
/>,
          ],
          "className": "product-table__header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Product Name",
              "className": "text-col",
            },
            "ref": null,
            "rendered": "Product Name",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "SKU",
              "className": "text-col",
            },
            "ref": null,
            "rendered": "SKU",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Amount (in Bits)",
              "className": "text-col",
            },
            "ref": null,
            "rendered": "Amount (in Bits)",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "In Development",
              "className": "select-col",
            },
            "ref": null,
            "rendered": "In Development",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Broadcast",
              "className": "select-col",
            },
            "ref": null,
            "rendered": "Broadcast",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "dirty-col",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "function",
        "props": Object {
          "handleValueChange": [Function],
          "product": Object {
            "amount": "1",
            "broadcast": "true",
            "displayName": "",
            "inDevelopment": "true",
            "sku": "",
            "validationErrors": Object {},
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="product-table__add-button"
              onClick={[Function]}
>
              Add Product
</button>,
            <button
              className="product-table__save-button"
              disabled={false}
              onClick={[Function]}
>
              Save All
</button>,
          ],
          "className": "product-table__buttons",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add Product",
              "className": "product-table__add-button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add Product",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Save All",
              "className": "product-table__save-button",
              "disabled": false,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Save All",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "",
          <div
            className="product-table__header"
>
            <div
                        className="text-col"
            >
                        Product Name
            </div>
            <div
                        className="text-col"
            >
                        SKU
            </div>
            <div
                        className="text-col"
            >
                        Amount (in Bits)
            </div>
            <div
                        className="select-col"
            >
                        In Development
            </div>
            <div
                        className="select-col"
            >
                        Broadcast
            </div>
            <div
                        className="dirty-col"
            />
</div>,
          Array [
            <Unknown
              handleValueChange={[Function]}
              product={
                            Object {
                                          "amount": "1",
                                          "broadcast": "true",
                                          "displayName": "",
                                          "inDevelopment": "true",
                                          "sku": "",
                                          "validationErrors": Object {},
                                        }
              }
/>,
          ],
          <div
            className="product-table__buttons"
>
            <button
                        className="product-table__add-button"
                        onClick={[Function]}
            >
                        Add Product
            </button>
            <button
                        className="product-table__save-button"
                        disabled={false}
                        onClick={[Function]}
            >
                        Save All
            </button>
</div>,
        ],
        "className": "product-table",
      },
      "ref": null,
      "rendered": Array [
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="text-col"
>
                Product Name
</div>,
              <div
                className="text-col"
>
                SKU
</div>,
              <div
                className="text-col"
>
                Amount (in Bits)
</div>,
              <div
                className="select-col"
>
                In Development
</div>,
              <div
                className="select-col"
>
                Broadcast
</div>,
              <div
                className="dirty-col"
/>,
            ],
            "className": "product-table__header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Product Name",
                "className": "text-col",
              },
              "ref": null,
              "rendered": "Product Name",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "SKU",
                "className": "text-col",
              },
              "ref": null,
              "rendered": "SKU",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Amount (in Bits)",
                "className": "text-col",
              },
              "ref": null,
              "rendered": "Amount (in Bits)",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "In Development",
                "className": "select-col",
              },
              "ref": null,
              "rendered": "In Development",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Broadcast",
                "className": "select-col",
              },
              "ref": null,
              "rendered": "Broadcast",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "dirty-col",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "function",
          "props": Object {
            "handleValueChange": [Function],
            "product": Object {
              "amount": "1",
              "broadcast": "true",
              "displayName": "",
              "inDevelopment": "true",
              "sku": "",
              "validationErrors": Object {},
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="product-table__add-button"
                onClick={[Function]}
>
                Add Product
</button>,
              <button
                className="product-table__save-button"
                disabled={false}
                onClick={[Function]}
>
                Save All
</button>,
            ],
            "className": "product-table__buttons",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add Product",
                "className": "product-table__add-button",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add Product",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Save All",
                "className": "product-table__save-button",
                "disabled": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Save All",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
